# AMR tools Benchmarking

- [Patric data set](#data)
    - [The procedures for data acquisition](#pro)
    - [Resulting species and antibiotics based on **strict** Quality Control](#strict)
    - [Resulting species and antibiotics based on **loose** Quality Control](#loose)
    - [Get IDs for each species and antibiotic combination](#u)
        - [Get the ID list directly from txt files](#files)
        - [Use /Patric_data_set/get_data.py](#getdata)
- [ResFinder](#p)
- [Multi-species](#m)


## <a name="data"></a>Patric data set
### <a name="pro"></a>**1. The procedures for data acquisition:**

- Download all the genomes with AMR phenotype meta-data generated by laboratory methods. This results in 99 species, including 67836 strains/genomes.

- Retain the species that contain larger than 500 strains. This results in 13 species.

**The Sequence ID for each species are in /Patric_data_set/metadata/model/id_$species_name.list, e.g. id_Pseudomonas_aeruginosa.list.**

- Qaulity control over 13 selected species: 

(1). Quality control criteria: strict

    - genome status : not Plasmid
    - genome_quality : Good
    - contigs <= 100
    - fine_consistency>= 97
    - coarse_consistency >= 98
    - checkm_completeness >= 98
    - checkm_contamination <= 2
    - |genome length - mean length| <= mean_length/20


(2). Quality control criteria: loose

    - genome status : not Plasmid
    - genome_quality : Good
    - contigs <=  100 or 75% Quartile (the one is chosen to include more strains, w.r.t. each species)
    - fine_consistency>= 97
    - coarse_consistency >= 98
    - checkm_completeness >= 98 and (missing value also accepted)
    - checkm_contamination <= 2  and  (missing value also accepted)
    - |genome length - mean length| <= mean_length/20
   

Strain number w.r.t. each species before and after Quality control:

![plot](./Patric_data_set/strain_n.png) 

- Retain strains with phenotype either R(resistant) or S(susceptible), i.e. strains with only I(intermediate)or none phenotype are excluded.

- Retain the species with at least 100 strains w.r.t.  each antibioticâ€™s each R and S phenotype classrespectively. 


### <a name="strict"></a>2. Resulting species and antibiotics based on **strict** Quality Control:

![plot](./Patric_data_set/species_strict.png) 



    
### <a name="loose"></a>3. Resulting species and antibiotics based on **loose** Quality Control:

![plot](./Patric_data_set/species_loose.png)



### <a name="u"></a> 4. Getting IDs for each species and antibiotic combination

**A. Get the ID list directly from txt files.**

A.1 Strict Quality Control: 

The Sequence ID for each species and antibiotic combination are in /Patric_data_set/metadata/model/strict/Data_${species_name}${antibiotic_name}.txt, e.g. Data_Escherichia_coli_amoxicillin.txt

A.2 Loose Quality Control:

The Sequence ID for each species and antibiotic combination are in /Patric_data_set/metadata/model/loose/Data_${species_name}${antibiotic_name}.txt, e.g. Data_Escherichia_coli_amoxicillin.txt

**B. Use /Patric_data_set/get_data.py**

2.1 Usage


```
cd /Patric_data_se
python get_data.py TODO
````

```
TODO
Providing additional functions of downsampling for unbalanced data set.

unbalance definition:

balance_ratio=(Number of strains in Susceptible)/(Number of strains in Resistance)

balance_ratio > 2 or balance_ratio < 0.5
```


2.2 Example

```

```



## <a name="p"></a> ResFinder

Bortolaia, Valeria, et al. "ResFinder 4.0 for predictions of phenotypes from genotypes." Journal of Antimicrobial Chemotherapy 75.12 (2020): 3491-3500.

https://bitbucket.org/genomicepidemiology/resfinder/src/master/



## <a name="m"></a> Multi-species
Nguyen, Marcus, et al. "Predicting antimicrobial resistance using conserved genes." PLoS computational biology 16.10 (2020): e1008319.

https://bitbucket.org/deaytan/data_preparation
https://bitbucket.org/deaytan/neural_networks/src/master/




